### DISCUSSION TITLE
Multi threads ADC read and saving it to a local file

### DISCUSSION THREAD

**Discussion by Alirezaaraby** [micropython/micropython#13486] [RP2040 / Pico]
Hi, i want to make an audio recorder with raspberry pi pico and max 9814. But the problem is when i want to save the recorded audio into a .wav file it loses some parts of the audio because it runs on one core and i want to write a code to handle the process of saving data on another core.

**Comment by peterhinch** [comment]
It could be difficult to make this work.

I would consider an alternative approach: a MEMS microphone with an I2S interface such as [this one](https://www.adafruit.com/product/3421). The RP2 has support for [i2s](https://docs.micropython.org/en/latest/rp2/quickref.html#i2s-bus) and there is an asynchronous API. It should therefore be possible to concurrently acquire data and save it using suitable buffering and `asyncio`.

This is something I intended to do, but to date I haven't found time.

**Reply by Alirezaaraby** [reply]
thanks for your help. I bought this https://www.adafruit.com/product/3421 microphone but there's no tutorial on the internet for connecting it to rpi pico. can you give me a hand on how I can wire it?

**Reply by ned-pcs** [reply]
Take a look at https://github.com/miketeachman/micropython-i2s-examples also.
And read the datasheet: https://www.knowles.com/docs/default-source/default-document-library/sph0645lm4h-1-datasheet.pdf
Note especially that BCLK must be 64x of LRCLK, so the microphone should be set up as a 32-bit stereo, but only alternate 32-bit samples will have actual data.
```
i2s = I2S(1, sck=Pin(0), ws=Pin(1), sd=Pin(2), mode=I2S.RX, bits=32, format=I2S.STEREO, rate=22050, ibuf=40000) # create I2S object
i2s.readinto(buf)          # fill buffer with audio samples from I2S device
```
  - BCLK on the microphone is connected to the `sck` pin on the Pico.
  - DOUT on the microphone is connected to the `sd` pin on the Pico.
  - LRCLK on the microphone is connected to the `ws` pin on the Pico.

**Reply by Alirezaaraby** [reply]
Thanks, it worked
